r/src/find_met_files.r

```diff
@@ -1,42 +1,50 @@
 #' find_met_files searches for meteorological data files
-#' @author Ben Fasoli
+#' @author Ben Fasoli & James Mineau
 #'
 #' Searches for available meteorological files matching the given strftime
 #' compatible file naming convention
 #'
 #' @param t_start time of simulation start
+#' @param n_hours number of hours to run each simulation; negative indicates
+#'   backward in time
+#' @param n_hours_per_met_file number of hours of meteorological data in each
+#'   met file
 #' @param met_file_format grep compatible file naming convention to identify
 #'   meteorological data files necessary for the timing of the simulation
 #'   indicated by \code{t_start} and \code{n_hours}
-#' @param n_hours number of hours to run each simulation; negative indicates
-#'   backward in time
 #' @param met_path directory to find meteorological data
 #'
 #' @import dplyr
 #' @export
 
-find_met_files <- function(t_start, met_file_format, n_hours, met_path) {
+find_met_files <- function(t_start, n_hours, n_hours_per_met_file,
+                           met_file_format, met_path) {
   require(dplyr)
-  
-  is_backward <- n_hours < 0
-  
-  # TODO: implement n_hours_per_met_file to better determine file names at
-  # varying time resolutions
-  request <- as.POSIXct(t_start, tz='UTC') %>%
-    c(. + c(1, -1, n_hours, is_backward * (n_hours - 5)) * 3600) %>%
```

Comment:
My brain isn't very good with R syntax anymore.

I think this adds a buffer (+1, -1 hours) around the time window [t_start, n_hours * n_hours_per_met_file] since there were some edge cases when simulations start or end at the boundary of one met file.

It looks like that buffer has been removed in this diff. Is that an expected change?

Reply:
The buffer is only around `t_start`, not the full window.

For example, lets set `t_start <- '2022-01-01'`
For the backward case:
```r
n_hours <- -24
is_backward <- n_hours < 0
as.POSIXct(t_start, tz='UTC') %>%
  c(. + c(1, -1, n_hours, is_backward * (n_hours - 5)) * 3600) %>%
  sort()
```
> "2021-12-30 19:00:00 UTC" "2021-12-31 00:00:00 UTC" "2021-12-31 23:00:00 UTC" "2022-01-01 00:00:00 UTC" "2022-01-01 01:00:00 UTC"

And the forward case:
```r
n_hours <- 24
is_backward <- n_hours < 0
as.POSIXct(t_start, tz='UTC') %>%
  c(. + c(1, -1, n_hours, is_backward * (n_hours - 5)) * 3600) %>%
  sort()
```
> "2021-12-31 23:00:00 UTC" "2022-01-01 00:00:00 UTC" "2022-01-01 00:00:00 UTC" "2022-01-01 01:00:00 UTC" "2022-01-02 00:00:00 UTC"

We only buffer -1 before `t_start`. For met files like hrrr which are 6-hourly resolution, we miss the first met file if `t_start` isn't 1 hour away from the file name.

Whereas my changes extend the met file bracket.
For the backward case:
```r
ts <- '2022-01-01'
n_hours <- -24
n_hours_per_met_file <- 6
is_backward <- as.numeric(n_hours < 0)
ib <- ifelse(is_backward, 1, -1)

seq(
    ts - (ib * as.difftime(abs(n_hours) + ib, units = 'hours')),
    ts + (ib * as.difftime(n_hours_per_met_file - is_backward, units = 'hours')),
    by = ib * 3600
  )
```
> [1] "2021-12-30 23:00:00 UTC" "2021-12-31 00:00:00 UTC" "2021-12-31 01:00:00 UTC" "2021-12-31 02:00:00 UTC"
 [5] "2021-12-31 03:00:00 UTC" "2021-12-31 04:00:00 UTC" "2021-12-31 05:00:00 UTC" "2021-12-31 06:00:00 UTC"
 [9] "2021-12-31 07:00:00 UTC" "2021-12-31 08:00:00 UTC" "2021-12-31 09:00:00 UTC" "2021-12-31 10:00:00 UTC"
[13] "2021-12-31 11:00:00 UTC" "2021-12-31 12:00:00 UTC" "2021-12-31 13:00:00 UTC" "2021-12-31 14:00:00 UTC"
[17] "2021-12-31 15:00:00 UTC" "2021-12-31 16:00:00 UTC" "2021-12-31 17:00:00 UTC" "2021-12-31 18:00:00 UTC"
[21] "2021-12-31 19:00:00 UTC" "2021-12-31 20:00:00 UTC" "2021-12-31 21:00:00 UTC" "2021-12-31 22:00:00 UTC"
[25] "2021-12-31 23:00:00 UTC" "2022-01-01 00:00:00 UTC" "2022-01-01 01:00:00 UTC" "2022-01-01 02:00:00 UTC"
[29] "2022-01-01 03:00:00 UTC" "2022-01-01 04:00:00 UTC" "2022-01-01 05:00:00 UTC

And the forward case:
```r
n_hours <- 24
is_backward <- as.numeric(n_hours < 0)
ib <- ifelse(is_backward, 1, -1)

seq(
    ts - (ib * as.difftime(abs(n_hours) + ib, units = 'hours')),
    ts + (ib * as.difftime(n_hours_per_met_file - is_backward, units = 'hours')),
    by = ib * 3600
  )
```
>  [1] "2021-12-31 18:00:00 UTC" "2021-12-31 19:00:00 UTC" "2021-12-31 20:00:00 UTC" "2021-12-31 21:00:00 UTC"
 [5] "2021-12-31 22:00:00 UTC" "2021-12-31 23:00:00 UTC" "2022-01-01 00:00:00 UTC" "2022-01-01 01:00:00 UTC"
 [9] "2022-01-01 02:00:00 UTC" "2022-01-01 03:00:00 UTC" "2022-01-01 04:00:00 UTC" "2022-01-01 05:00:00 UTC"
[13] "2022-01-01 06:00:00 UTC" "2022-01-01 07:00:00 UTC" "2022-01-01 08:00:00 UTC" "2022-01-01 09:00:00 UTC"
[17] "2022-01-01 10:00:00 UTC" "2022-01-01 11:00:00 UTC" "2022-01-01 12:00:00 UTC" "2022-01-01 13:00:00 UTC"
[21] "2022-01-01 14:00:00 UTC" "2022-01-01 15:00:00 UTC" "2022-01-01 16:00:00 UTC" "2022-01-01 17:00:00 UTC"
[25] "2022-01-01 18:00:00 UTC" "2022-01-01 19:00:00 UTC" "2022-01-01 20:00:00 UTC" "2022-01-01 21:00:00 UTC"
[29] "2022-01-01 22:00:00 UTC" "2022-01-01 23:00:00 UTC"


See the diagrams for a graphic representation:
![image](https://github.com/user-attachments/assets/cc372890-d960-40c5-adca-8e4ee0876680)
![image](https://github.com/user-attachments/assets/f69796e6-2fef-4e88-8a75-70d58e21bbc4)

Reply:
I'm not sure if `hycs_std` needs initial conditions, but looking at my diagrams again, if we don't need initial conditions, then I think the met bracket (red) could be just `n_hours_per_met_file - 1`?

Though now Im wondering if we need to add `(n_hours_per_met_file - 1`) to `abs(n_hours)` for the simulation bracket (blue)? Otherwise we could miss the end (chronologically start) of the backwards simulations

Reply:
I think this might be the way to go actually:

```r
ts <- as.POSIXct(t_start, tz = 'UTC')
is_backward <- n_hours < 0
met_bracket <- n_hours_per_met_file - 1  # ts can be in the middle of a met file

if (is_backward) {
  request <- seq(
    ts - as.difftime(abs(n_hours) + is_backward * (met_bracket), units = 'hours'),
    ts,
    by = 3600
  )
} else {
  request <- seq(
    ts - as.difftime(met_bracket, units = 'hours'),
    ts + as.difftime(n_hours, units = 'hours'),
    by = 3600
  )
}


![image](https://github.com/user-attachments/assets/8cf18384-44fe-45ed-85a1-7fbe2ef493f8)

Reply:
I think that makes sense 👍